File: raspberry-pi-ios-client/RaspberryPiClient/Data/Infrastructure/Foxglove/Binary/FoxgloveBinaryMessageEncoder.swift
//
//  FoxgloveBinaryMessageEncoder.swift
//  RaspberryPiClient
//
//  Created by Shusuke Ota on 2025/03/20.
//

import Foundation

/// クライアント → サーバーへのバイナリメッセージエンコーダ
/// Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#binary-messages
struct FoxgloveBinaryMessageEncoder {
    /// Client Message Data（opcode 0x01）をエンコードする
    /// - Parameters:
    ///   - channelId: Client Advertise で登録したチャンネル ID
    ///   - payload: メッセージペイロード（JSON エンコード済みなど）
    /// - Returns: 送信用バイナリデータ
    static func encodeClientMessageData(channelId: UInt32, payload: Data) -> Data {
        var data = Data()
        data.append(FoxgloveClientBinaryOpcode.clientMessageData.rawValue)
        data.append(contentsOf: withUnsafeBytes(of: channelId.littleEndian) { Array($0) })
        data.append(payload)
        return data
    }

    /// Service Call Request（opcode 0x02）をエンコードする
    /// - Parameters:
    ///   - serviceId: Advertise Services で取得したサービス ID
    ///   - callId: レスポンスと対応付けるための一意な ID
    ///   - encoding: エンコーディング名（例: "json"）
    ///   - payload: リクエストペイロード
    /// - Returns: 送信用バイナリデータ
    static func encodeServiceCallRequest(serviceId: UInt32, callId: UInt32, encoding: String, payload: Data) -> Data {
        guard let encodingData = encoding.data(using: .utf8) else {
            return Data()
        }
        let encodingLength = UInt32(encodingData.count)

        var data = Data()
        data.append(FoxgloveClientBinaryOpcode.serviceCallRequest.rawValue)
        data.append(contentsOf: withUnsafeBytes(of: serviceId.littleEndian) { Array($0) })
        data.append(contentsOf: withUnsafeBytes(of: callId.littleEndian) { Array($0) })
        data.append(contentsOf: withUnsafeBytes(of: encodingLength.littleEndian) { Array($0) })
        data.append(encodingData)
        data.append(payload)
        return data
    }
}

/// クライアント → サーバーのバイナリメッセージ opcode（本ファイル内でのみ使用）
private enum FoxgloveClientBinaryOpcode: UInt8 {
    case clientMessageData = 0x01
    case serviceCallRequest = 0x02
}

File: raspberry-pi-android-client/app/src/main/java/com/shusuke/raspberry_pi_client/data/infrastructure/foxglove/binary/FoxgloveBinaryMessageEncoder.kt
package com.shusuke.raspberry_pi_client.data.infrastructure.foxglove.binary

import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * クライアント → サーバーへのバイナリメッセージエンコーダ
 *
 * Ref: https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#binary-messages
 */
object FoxgloveBinaryMessageEncoder {

    private const val OPCODE_CLIENT_MESSAGE_DATA: Byte = 0x01
    private const val OPCODE_SERVICE_CALL_REQUEST: Byte = 0x02

    /**
     * Client Message Data（opcode 0x01）をエンコードする
     *
     * @param channelId Client Advertise で登録したチャンネル ID
     * @param payload メッセージペイロード（JSON エンコード済みなど）
     * @return 送信用バイナリデータ
     */
    fun encodeClientMessageData(channelId: UInt, payload: ByteArray): ByteArray {
        val buffer = ByteBuffer.allocate(1 + 4 + payload.size)
            .order(ByteOrder.LITTLE_ENDIAN)
        buffer.put(OPCODE_CLIENT_MESSAGE_DATA)
        buffer.putInt(channelId.toInt())
        buffer.put(payload)
        return buffer.array()
    }

    /**
     * Service Call Request（opcode 0x02）をエンコードする
     *
     * @param serviceId Advertise Services で取得したサービス ID
     * @param callId レスポンスと対応付けるための一意な ID
     * @param encoding エンコーディング名（例: "json"）
     * @param payload リクエストペイロード
     * @return 送信用バイナリデータ
     */
    fun encodeServiceCallRequest(
        serviceId: UInt,
        callId: UInt,
        encoding: String,
        payload: ByteArray,
    ): ByteArray {
        val encodingBytes = encoding.encodeToByteArray()
        val encodingLength = encodingBytes.size
        val buffer = ByteBuffer.allocate(1 + 4 + 4 + 4 + encodingLength + payload.size)
            .order(ByteOrder.LITTLE_ENDIAN)
        buffer.put(OPCODE_SERVICE_CALL_REQUEST)
        buffer.putInt(serviceId.toInt())
        buffer.putInt(callId.toInt())
        buffer.putInt(encodingLength)
        buffer.put(encodingBytes)
        buffer.put(payload)
        return buffer.array()
    }
}
